src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -539,9 +539,12 @@ void QgsLayerTreeModel::refreshLayerLegend( QgsLayerTreeLayer *nodeLayer )
 
   // update children
   int oldNodeCount = rowCount( idx );
-  beginRemoveRows( idx, 0, std::max( oldNodeCount - 1, 0 ) );
-  removeLegendFromLayer( nodeLayer );
-  endRemoveRows();
+  if ( oldNodeCount > 0 )
+  {
+    beginRemoveRows( idx, 0, std::max( oldNodeCount - 1, 0 ) );
```

Comment:
```suggestion
    beginRemoveRows( idx, 0, oldNodeCount );
```

This check should no longer be required, no?

Reply:
Maybe it is dangerous to BeginRemoveRows that doesn't exist, no ? First solution avoid calling this method.

Reply:
This will only trigger a change if oldNodeCount <= 0 and has no effect for any oldNodeCount > 0, so basically it looks redundant with your check.

Reply:
OK, I've got the origin of my application crash. When I am doing some test with QgsMeshlayerProperties, sometime the mesh layer doesn't have legend (when only the mesh is visible : "show contours" unchecked). 
When properties are edited after "Apply" or "Ok" the oldNodeCount variable, in the refreshLayerLegend, method is 0. 

The beginRemoveRows method of the QAbtractItemModel class :


```
void QAbstractItemModel::beginRemoveRows(const QModelIndex &parent, int first, int last)
{
    Q_ASSERT(first >= 0);
    Q_ASSERT(last >= first);
    Q_ASSERT(last < rowCount(parent));
```

If last=0 and rowCount(parent) =0--> the program is stopped because I am in debug mode ...

The behaviour is not the same with my check, so it is not redundant. I don't know if this check is useful for the Qgis App, but it seems it could avoid the ASSERT being false.

Reply:
> When properties are edited after "Apply" or "Ok" the oldNodeCount variable, in the refreshLayerLegend, method is 0.

Sorry, it must be me standing on the wire.
Can you help me understand how line 544 is reached with oldNodeCount = 0? I can't help to think that this will be guarded on line 542.

Reply:
You are right, forgive me for the misunderstanding. Sometime, my understanding of english has some bugs ...
The max check is of course no longer be required.